### PR TITLE
[WIP] add kotlin allopen transformation

### PR DESCRIPTION
**Pull request by RotBolt**
cc @evanchooly @geoand 

Currently it is able to transform only classes with annotation
 but if a class (which is to be opened) extends a class (which also needs to be opened) than it fails

For example `class NativeImageTest : GreetingResourceTest`  then it fails 
probably because GreetingResourceTest has not been opened by transformer yet

### CODE REVIEW

**Comment by RotBolt** [comment]
Also one thing It came to mind 
to also check for super class and perform transformation for that too

**Comment by geoand** [comment]
This is a good start, but I think we need to have Quarkus automatically figure out which (Kotlin) classes need to be opened and not have this be driven by user config.
We should always know when a class needs to be opened (if it's a Hibernate entity or a JAX-RS resource for example).
How best to do this is up for discussion

**Comment by RotBolt** [comment]
Might be wrong path way to do

If we put the all the annotations for which we want the classes to be ooen. I guess it would solve the issue 

A better way to do would 
- check the dependencies (if includes hibernate or resteasy) 
- enable annotations for those dependencies
- perform transformation

**Comment by geoand** [comment]
There are 2 ways of figuring out what needs to be opened:

1) Each extension publishes this data (essentially pushes a build item)
2) The Kotlin extension has all the lookup logic.

Obviously 1) is better, but it requires changes to more extensions

**Comment by RotBolt** [comment]
Even if each extension pushes its own buildItem that build item has to be injected in the processor AFAIU.
So wouldn't this be kinda similar to "all lookup logic in kotlin" ?

**Comment by geoand** [comment]
The difference lies in where the logic of determining which classes need to be opened lies.

Also, one could think that making a class open (or non-final in Java terms) is a more general concern, not limited to Kotlin.

**Comment by evanchooly** [comment]
We might just be overthinking it, too.  Java is not closed by default where kotlin is.  We could simply apply this transformation to all Kotlin classes (check the for the kotlin metadata annotation) and all the get/set/is methods on the classes.  This would treat the cases where a Java dev marks a class as final as a user error to be cleaned up by the dev rather than silently sweep it under a rug.

**Comment by geoand** [comment]
Yeah, that's definitely an option as well.

The only problem I foresee with that approach is that applying this transformation to all Kotlin classes has the potential to slow down the build.

File: extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/AllOpenConfig.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "allopen")
+public class AllOpenConfig {
+
+    public final List<String> defaultAnnotations = Arrays.asList(
```
Comment (evanchooly) [inline extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/AllOpenConfig.java:12]: use DotName instead

File: extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java [hunk 2]
```diff
@@ -37,4 +52,60 @@ void registerKotlinJacksonModule(BuildProducer<ClassPathJacksonModuleBuildItem>
...
+        Collection<ClassInfo> classes = applicationIndexBuildItem.getIndex().getKnownClasses();
+
+        ArrayList<String> classesToTransform = new ArrayList<>();
+        for (ClassInfo classInfo : classes) {
+            for (List<AnnotationInstance> annotationInstances : classInfo.annotations().values()) {
+
+                for (AnnotationInstance annotationInstance : annotationInstances) {
+//                    System.out.println("Name : " + annotationInstance.name());
+//                    System.out.println("Value " + annotationInstance.value());
+//                    System.out.println("Target " + annotationInstance.target().toString());
```
Comment (evanchooly) [inline extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java:71]: debug should come out

File: extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java [hunk 3]
```diff
@@ -37,4 +52,60 @@ void registerKotlinJacksonModule(BuildProducer<ClassPathJacksonModuleBuildItem>
...
+
+                    for (String defaultAnnotation : allOpenConfig.defaultAnnotations) {
+                        if (annotationInstance.name().toString().equals(defaultAnnotation)) {
+                            classesToTransform.add(classInfo.name().toString());
```
Comment (evanchooly) [inline extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java:75]: using DotNames a simple .equals would suffice

File: extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java [hunk 4]
```diff
@@ -37,4 +52,60 @@ void registerKotlinJacksonModule(BuildProducer<ClassPathJacksonModuleBuildItem>
...
+                        }
+                    }
+                    for (String customAnnotation : allOpenConfig.fqCustomAnnotations) {
+                        if (annotationInstance.name().toString().equals(customAnnotation)) {
```
Comment (evanchooly) [inline extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java:79]: same

File: extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java [hunk 5]
```diff
@@ -37,4 +52,60 @@ void registerKotlinJacksonModule(BuildProducer<ClassPathJacksonModuleBuildItem>
...
+                        @Override
+                        public ClassVisitor apply(String s, ClassVisitor classVisitor) {
+                            ClassVisitor cv = new ClassVisitor(Gizmo.ASM_API_VERSION, classVisitor) {
+                                @Override
+                                public void visit(int version, int access, String name, String signature, String superName,
+                                                  String[] interfaces) {
+                                    super.visit(version, access, name, signature, superName, interfaces);
+                                    System.out.println("ByteCodeTransform " + name + " access " + access);
+                                    access = access ^ Opcodes.ACC_FINAL;
+                                    System.out.println("ACC_FINAL " + Opcodes.ACC_FINAL + " Access" + access);
```
Comment (evanchooly) [inline extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java:101]: debug

**Review by evanchooly — CHANGES_REQUESTED** [review]

**Comment by evanchooly** [comment]
if we wanted to generalize it, we could add a, say, `NonFinalBuildItem` that indicates an annotation that requires non-final classes and methods.  We could couple that with the kotlin metadata annotation to speed things up.

At that point the question becomes do we just always unmask the final for java classes, too?  It'd be the "neighborly" thing to do but i'm loathe to automatically, silently undo conscious decisions on the part of the Java developer.  We could, perhaps, just apply the filter to kotlin and java classes alike and log a `warning` for non-kotlin classes to direct the java devs to remove those final keywords.  This would mean each extension would have to publish that new build item but it would future proof this transformer against new extensions/annotations.

**Comment by geoand** [comment]
> if we wanted to generalize it, we could add a, say, `NonFinalBuildItem` that indicates an annotation that requires non-final classes and methods.  We could couple that with the kotlin metadata annotation to speed things up.
> 
 Exactly what I was thinking

**Comment by geoand** [comment]
@evanchooly yeah, I'm not sure what the best solution is, just pondering options

**Comment by RotBolt** [comment]
> if we wanted to generalize it, we could add a, say, `NonFinalBuildItem` that indicates an annotation that requires non-final classes and methods.  We could couple that with the kotlin metadata annotation to speed things up.


Would the extension push these `NonFinalBuildItem` ? And we would just unmask the final for the classes which has `kotlin metadata` and these `NonFinalBuildItem` ?

**Comment by evanchooly** [comment]
Well, if the extensions were publishing these theoretical `NonFinalBuildItem`s we would simply mask away final on any classes so marked regardless of the kotlin metadata annotation.  We could optionally log that we were masking away final modifiers and either use the kotlin metadata annotation to limit that or not.

**Comment by evanchooly** [comment]
adding to epic  #10377

**Comment by gsmet** [comment]
@evanchooly is it still needed with all the work you did? I would be in favor of closing this one and revisit in another PR if need be. Could you confirm?

**Comment by evanchooly** [comment]
I agree.  so far quarkus makes what changes it needs and the fewer assumptions we can make about external configurations the simpler we can make things internally.

**Comment by gastaldi** [comment]
Closing as described above

**Pull request closed without merging**